### PR TITLE
The new RM implementation

### PR DESCRIPTION
**Pull request by ragrus-nbl**
The new RM:
1) Is fully independent from MessageManager.
2) Uses SPIFlashLogger and ConnectionManager.
3) Consists of two parts (classes): base and extended (inherited from the base part).
4) Every message has "importance" parameter which will be specified by user in the send() method. "Importance" can be: 0, 1 and 2.
5) If importance = 0, the message will not be persisted and will not be resent in case of failure. Just one try and the onFail() will be called.
6) If importance = 1, the message will be automatically persisted in case of failure (with no call of the onFail()). Then it will be "replayed" once connection is reestablished.
7) If importance = 2, the message will be instantly persisted. In case of failure it will be "replayed" once connection is reestablished.
8) The base part supports only importance = 0 and is intended to be used on imp agent.
9) The extended version supports all values of importance and is intended to be used on imp device when persistence is needed.
10) No local (message-specific) callbacks. Because all local callbacks will be lost once message is persisted (a function cannot be saved to flash) or device rebooted. The main point of local callbacks is the ability of usage of local context. But if user has to re-set their callbacks (on "replaying" messages from storage or reboot), they will have their context already been lost.
11) There are 3 global callbacks for outbound messages: onAck(), onFail(), confirmResend() (available only in the extended part).
12) For inbound messages there are on() and defaultOn() callbacks.
13) confirmResend() callback is called during the "replaying" process - when the connection has been reestablished and the lib suggests to user to resend some (or all) of pending messages persisted on the flash. This callback is called for every message from the flash so that user is able to "evaluate" it and send/not send. confirmResend() user callback must return bool value: true to resend the message and false otherwise.
14) If confirmResend() returns false, the onFail() callback is called for the message.
15) If the lib has to delete some message from flash memory (memory is full), the onFail() callback is called for that message.
16) Instead of replies, customized acknowledges (customAck's) are allowed.
17) An opportunity to manage acks is given to applications. An application will be able to: send an empty ack, send an ack with some data, not send any ack.
18) An ack can be sent synchronously and asynchronously as well.
19) customAck is a kind of non-guaranteed reply which is assumed to be sent within the specified timeout. The sender side decides the sending fails (and may resend the same msg) if no ack is received within that timeout.
IMG_20190305_145145

### CODE REVIEW

File: ReplayMessenger.extended.lib.nut [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+        }
+    }
+
+    // Calls the error handler if set for the message with the specified error
+    function _onSendFail(msg, error) {
+        local id = msg.payload.id;
+        _log("Failed to send or haven't received Ack. Id: " + id + " Reason: " + error);
+
+        // If importance is RM_IMPORTANCE_HIGH, persist the message if not yet
+        if(msg._importance == RM_IMPORTANCE_HIGH && !_isMsgPersisted(msg)) {
```
Comment (ppetrosh) [inline ReplayMessenger.extended.lib.nut]: Can we please make the code conformant with the coding style requirements.
"if("->"if (", etc.

Here and everywhere, I won't post it for every single case I find.

Comment (ragrus-nbl) [reply ReplayMessenger.extended.lib.nut]: Fixed!

File: ReplayMessenger.base.lib.nut [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    // The primary purpose is rate measurements, where the above
+    // limitation is not critical.
+    function _monotonicMillis() {
+        return _isAgent() ? time() * 1000 + date().usec / 1000 : hardware.millis();
+    }
+
+    function _typeof() {
+        return "ReplayMessengerBase";
+    }
+}
```
Comment (ppetrosh) [inline ReplayMessenger.base.lib.nut]: It's better to have empty lines at the end. I remember there were some edge cases triggered by this in Builder and on the impCentral sides )

**Review by ppetrosh — APPROVED** [review]

**Comment by ppetrosh** [comment]
We should think about how to better name the classes. For example, Extended is not used on the agent side, while the Base can/should be used on both: agent and device sides. Something to think about in the future

**Pull request merged**